src/object_based/extent_htm.cpp

```diff
@@ -0,0 +1,352 @@
+    // triangle (even though it's possible for a point to be in multiple
+    // triangles because of tolerance)
+    //
+    // assign_points_to_tris works recursively writing into points and
+    // point_pvs at next_point, in depth-first order
+    std::vector<bool> assigned(input_points.cols(), false);
+    size_t next_point = 0;
+    for (size_t base_tri : htm.initial_tri_ids)
+      assign_points_to_tris(input_points, input_pvs, assigned, next_point,
+                            base_tri);
+
+    ear_assert(next_point == (size_t)points.cols(), "missed a point");
+    for (bool tri_assigned : assigned)
+      ear_assert(tri_assigned, "point not assigned to any tri");
+  }
+
+  void SpreadingPannerHTM::assign_points_to_tris(
+      const Eigen::Matrix<double, 3, Eigen::Dynamic> &input_points,
+      const Eigen::MatrixXd &input_pvs, std::vector<bool> &assigned,
+      size_t &next_point, size_t tri_id) {
+    size_t start = next_point;
+    if (!htm.tris[tri_id].has_next) {
+      // if this is a leaf triangle, assign any un-assigned points to it
+      HTM::Point &p1 = htm.points[htm.tris[tri_id].points[0]];
+      HTM::Point &p2 = htm.points[htm.tris[tri_id].points[1]];
+      HTM::Point &p3 = htm.points[htm.tris[tri_id].points[2]];
+
+      for (size_t i = 0; i < (size_t)input_points.cols(); i++) {
+        if (!assigned[i] &&
+            pointInsideSphericalTriangle(p1, p2, p3, input_points.col(i))) {
+          assigned[i] = true;
+          points.col(next_point) = input_points.col(i);
+          point_pvs.col(next_point) = input_pvs.col(i);
+          next_point++;
+          tri_pvs.col(tri_id) += input_pvs.col(i);
+        }
+      }
+    } else {
+      // otherwise, assign points to the subdivisions of this triangle
+      for (size_t next_tri : htm.tris[tri_id].next_tris) {
+        assign_points_to_tris(input_points, input_pvs, assigned, next_point,
+                              next_tri);
+        tri_pvs.col(tri_id) += tri_pvs.col(next_tri);
+      }
+    }
+    size_t end = next_point;
+
+    tri_points_ptrs[tri_id] = {start, end};
+  }
+
+  Eigen::VectorXd SpreadingPannerHTM::panningValuesForWeight(
+      const WeightingFunction &weightFunc) {
+    point_cache.clear();
+    inside_edge_cache.clear();
+    outside_edge_cache.clear();
+
+    WeightingFunctionEdges edges(weightFunc);
+
+    Eigen::VectorXd totalPv = Eigen::VectorXd::Zero(point_pvs.rows());
+    for (size_t base_tri : htm.initial_tri_ids)
+      integrate(weightFunc, edges, base_tri, totalPv);
+
+    totalPv /= totalPv.norm();
+    return totalPv;
+  }
+
+  void SpreadingPannerHTM::integrate(const WeightingFunction &wf,
+                                     const WeightingFunctionEdges &edges,
+                                     size_t tri_id, Eigen::VectorXd &pvs) {
+    HTM::CTri &tri = htm.tris[tri_id];
+    bool inside, outside;
+    std::tie(inside, outside) = edges.spherical_tri_inside_outside(
```

Comment:
Just adding to the tuple comment above, I think std::tie works by constructing a temporary tuple of <T1&, T2&> then assigning to it, where as if you return a struct it probably just gets constructed in place or optimised out entirely. Usually doesn't matter one bit, but as you were eliminating temporaries in the earlier commit thought it made sense to comment.